File: 01_Bernoulli/bernoulli.py
import numpy as np
import sys
from random import random

def main(n):
    d = gen_data(n)
    mu_ml = mle(d)
    print("mle: ", mu_ml)
    bayes = Beta(2,2)
    bayes.update(d)
    print(bayes)
    print("optimal: ", bayes.find_optimal())

def gen_data(n):
    return np.random.randint(2, size=n)

def mle(d):
    return np.sum(d) / len(d)

class Beta:
    def __init__(self, a, b):
        self.a = a
        self.b = b

    def update(self, d):
        m = (d == 1).sum()
        l = len(d) - m
        self.a += m
        self.b += l

    def __str__(self):
        return "Beta(" + str(self.a) + ", " + str(self.b) + ")"

    def find_optimal(self):
        return (self.a - 1) / (self.a + self.b - 2)

if __name__=="__main__":
    main(int(sys.argv[1]))

File: 01_Bernoulli/bernoulli.jl
using Pipe

function main(args)
    n = @pipe args[1] |> parse(Int, _)
    d = gen_data(n)
    μ_ml = mle(d)
    println("mle: ", μ_ml)
    bayes = Beta(2, 2)
    @pipe bayes |> update(_, d)
    println(bayes)
    println("optimal: ", bayes |> find_optimal)
end

function gen_data(n)
    rand((0, 1), n)
end

function mle(d::T) where {S <: Number, T <: AbstractArray{S}}
    sum(d) / length(d)
end

mutable struct Beta{T}
    a::T
    b::T
end

function update(β::Beta{T}, d) where T <: Number
    m = count(==(1), d)
    l = length(d) - m
    β.a += m
    β.b += l
end

function find_optimal(β::Beta{T}) where T <: Number
    (β.a - 1) / (β.a + β.b - 2)
end

main(ARGS)
